hpy/devel/include/hpy/hpytype.h

```diff
@@ -76,43 +85,66 @@ typedef struct {
 
        HPyType_HELPERS(PointObject)
 
+   It is also possible to inherit from some built-in types. The list of
+   available built-in base types is given in enum `HPyTupe_BuiltinShape`.
+   In case you want to inherit from one of those, it is necessary to specify
+   the base built-in type in the `HPyType_HELPERS` macro. Here is an example
+   for a pure HPy custom type inheriting from a built-in type 'tuple':
+
+       HPyType_HELPERS(PointObject, HPyType_BuiltinShape_Tuple)
+
    This would generate the following:
 
    * `PointObject * PointObject_AsStruct(HPyContext *ctx, HPy h)`: a static
      inline function that uses HPy_AsStruct to return the PointObject struct
      associated with a given handle. The behaviour is undefined if `h`
      is associated with an object that is not an instance of PointObject.
 
-   * `PointObject_IS_LEGACY`: an enum value set to 0 so that in the
-     HPyType_Spec for PointObject one can write
-     `.legacy = PointObject_IS_LEGACY` and not have to remember to
-     update the spec when the helpers used changes.
+   * `PointObject_SHAPE`: an enum value set to 0 so that in the HPyType_Spec
+     for PointObject one can write `.builtin_shape = PointObject_SHAPE` and not
+     have to remember to update the spec when the helpers used changes.
 
    Example for a legacy custom type:
 
        HPyType_LEGACY_HELPERS(PointObject)
 
    This would generate the same functions and constants as above, except:
 
-   * `HPy_AsStructLegacy` is used instead of `HPy_AsStruct`.
+   * `_HPy_AsStruct_Legacy` is used instead of `_HPy_AsStruct_Object`.
 
-   * `PointObject_IS_LEGACY` is set to 1.
+   * `PointObject_SHAPE` is set to `HPyType_BuiltinShape_Legacy`.
 */
 
+/* Just a few helper macros to select the appropriate argument from the
+   variadic-macro's arguments. */
+
+#define HPyType_BUILTIN_HELPERS(TYPE, SHAPE) \
+    _HPyType_GENERIC_HELPERS(TYPE, SHAPE)
+
 #define HPyType_HELPERS(TYPE) \
-    _HPyType_GENERIC_HELPERS(TYPE, HPy_AsStruct, 0)
+    _HPyType_GENERIC_HELPERS(TYPE, HPyType_BuiltinShape_Object)
 
 #define HPyType_LEGACY_HELPERS(TYPE) \
-    _HPyType_GENERIC_HELPERS(TYPE, HPy_AsStructLegacy, 1)
-
-#define _HPyType_GENERIC_HELPERS(TYPE, CAST, IS_LEGACY)              \
-                                                                     \
-enum { TYPE##_IS_LEGACY = IS_LEGACY };                               \
-                                                                     \
-HPyAPI_UNUSED TYPE *                                                 \
-TYPE##_AsStruct(HPyContext *ctx, HPy h)                              \
-{                                                                    \
-    return (TYPE*) CAST(ctx, h);                                     \
+    _HPyType_GENERIC_HELPERS(TYPE, HPyType_BuiltinShape_Legacy)
+
```

Comment:
I wonder whether we can improve/simplify the usage and have only `HPyType_HELPERS`, whose first param is optional. I.e.:
```c
HPyType_HELPERS(PointObject); // common case
HPyType_HELPERS(PointObject, HPyType_BuiltinShape_Object); // equivalent to the above
HPyType_HELPERS(PointObject, HPyType_BuiltinShape_Legacy);
HPyType_HELPERS(PointObject, HPyType_BuiltinShape_Tuple);
```

Reply:
I think it is possible and I tried to do that but it became unreadable so I gave up on it.
The problem is: we cannot have simply two macro definitions like:
```
#define HPyType_HELPERS(TYPE)
#define HPyType_HELPERS(TYPE, SHAPE)
```
So, we need to do something like:
```
#define HPyType_HELPERS(TYPE, ...)
```
If you do it this way, you quickly run into problems if `__VA_ARGS__` is empty. In particular, if you pass them on and have a preceding comma like:
```
DO_SOMETHING(TYPE, __VA_ARGS__)
```
So, you need to do
```
#define HPyType_HELPERS(...)
```
such that `__VA_ARGS__` is never empty but then selecting the single arguments out of the `__VA_ARGS__` is nasty.

Long story short: it is possible (I think) but the macro definition becomes ugly.
I'll give it another try ...

Reply:
As promised, I gave it another try and it works well.
The only drawback is that we need to expose more helper macros that like `_HPyType_HELPER_DEFINE_SHAPE`. But I'm not opposed to that.

That means: I've removed `HPyType_BUILTIN_HELPERS` and you can do:
```
HPyType_HELPERS(SomeObject)
HPyType_HELPERS(SomeObject, HPyType_BuiltinShape_Something)
```